File: stdlib/js/hull/tests/test_htmx_inline_edit.js
// test_htmx_inline_edit.js. Tests for hull:web:htmx:inline-edit.

import { inlineEdit } from "hull:web:htmx:inline-edit";

let pass = 0, fail = 0;
function test(name, fn) {
    try { fn(); pass++; }
    catch (e) { fail++; console.log("FAIL: " + name + ": " + e.message); }
}
function assertMatch(s, sub, msg) {
    if (s.indexOf(sub) < 0) throw new Error((msg || "") + " expected '" + sub + "' in '" + s + "'");
}
function assertNoMatch(s, sub, msg) {
    if (s.indexOf(sub) >= 0) throw new Error((msg || "") + " did not expect '" + sub + "' in '" + s + "'");
}

// ── cell() ──────────────────────────────────────────────────────────

test("cell emits a span with the value", () => {
    const s = inlineEdit.cell({ value: "Conveyor #3", editUrl: "/x/edit" });
    assertMatch(s, "<span");
    assertMatch(s, "Conveyor #3");
});

test("cell uses outerHTML swap", () => {
    assertMatch(inlineEdit.cell({ value: "x", editUrl: "/x/edit" }), 'hx-swap="outerHTML"');
});

test("cell is keyboard-activatable", () => {
    const s = inlineEdit.cell({ value: "x", editUrl: "/x/edit" });
    assertMatch(s, 'role="button"');
    assertMatch(s, 'tabindex="0"');
});

test("cell triggers on click, Enter, Space", () => {
    const s = inlineEdit.cell({ value: "x", editUrl: "/x/edit" });
    assertMatch(s, "click");
    assertMatch(s, "Enter");
    assertMatch(s, "key==' '");
});

test("cell emits hx-get pointing at editUrl", () => {
    assertMatch(
        inlineEdit.cell({ value: "x", editUrl: "/assets/42/name/edit" }),
        'hx-get="/assets/42/name/edit"',
    );
});

test("cell default label is Edit", () => {
    const s = inlineEdit.cell({ value: "x", editUrl: "/x/edit" });
    assertMatch(s, 'aria-label="Edit"');
    assertMatch(s, 'title="Edit"');
});

test("cell custom label", () => {
    assertMatch(
        inlineEdit.cell({ value: "x", editUrl: "/x/edit", label: "Edit asset name" }),
        'aria-label="Edit asset name"',
    );
});

test("cell escapes value", () => {
    const s = inlineEdit.cell({ value: "<script>alert(1)</script>", editUrl: "/x/edit" });
    assertNoMatch(s, "<script>");
    assertMatch(s, "&lt;script&gt;");
});

test("cell escapes editUrl", () => {
    assertMatch(
        inlineEdit.cell({ value: "x", editUrl: '/x?a="b"' }),
        'hx-get="/x?a=&quot;b&quot;"',
    );
});

// ── editor() ────────────────────────────────────────────────────────

test("editor emits a form with hx-patch", () => {
    const s = inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" });
    assertMatch(s, "<form");
    assertMatch(s, 'hx-patch="/x"');
});

test("editor uses outerHTML swap targeting itself", () => {
    const s = inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" });
    assertMatch(s, 'hx-target="this"');
    assertMatch(s, 'hx-swap="outerHTML"');
});

test("editor input has default name=value", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" }),
        'name="value"',
    );
});

test("editor input takes custom name", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view", name: "title" }),
        'name="title"',
    );
});

test("editor input pre-fills value", () => {
    assertMatch(
        inlineEdit.editor({ value: "Conveyor #3", saveUrl: "/x", cancelUrl: "/x/view" }),
        'value="Conveyor #3"',
    );
});

test("editor default labels are Save / Cancel", () => {
    const s = inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" });
    assertMatch(s, ">Save</button>");
    assertMatch(s, ">Cancel</button>");
});

test("editor custom button labels", () => {
    const s = inlineEdit.editor({
        value: "x", saveUrl: "/x", cancelUrl: "/x/view",
        saveLabel: "Update", cancelLabel: "Keep",
    });
    assertMatch(s, ">Update</button>");
    assertMatch(s, ">Keep</button>");
});

test("editor Cancel button points at cancelUrl", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" }),
        'hx-get="/x/view"',
    );
});

test("editor wires Esc-to-cancel", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" }),
        "Escape",
    );
});

test("editor escapes value (XSS in pre-fill)", () => {
    const s = inlineEdit.editor({
        value: '"><script>alert(1)</script>',
        saveUrl: "/x", cancelUrl: "/x/view",
    });
    assertNoMatch(s, "<script>");
    assertNoMatch(s, '"><script>');
    assertMatch(s, 'value="&quot;&gt;');
});

test("editor input has aria-label (default)", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view" }),
        'aria-label="Edit value"',
    );
});

test("editor custom aria-label", () => {
    assertMatch(
        inlineEdit.editor({ value: "x", saveUrl: "/x", cancelUrl: "/x/view", label: "Asset name" }),
        'aria-label="Asset name"',
    );
});

console.log(`\n${pass}/${pass + fail} inline-edit tests passed`);
if (fail > 0) throw new Error("inline-edit tests failed");

File: stdlib/lua/hull/tests/test_htmx_inline_edit.lua
-- test_htmx_inline_edit.lua. Tests for hull.web.htmx.inline-edit.

local inline_edit = require('hull.web.htmx.inline-edit')

local pass, fail = 0, 0
local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then pass = pass + 1
    else fail = fail + 1; print("FAIL: " .. name .. ": " .. tostring(err)) end
end
local function assert_match(s, sub, msg)
    if not s:find(sub, 1, true) then
        error((msg or "") .. " expected '" .. sub .. "' in '" .. s .. "'")
    end
end
local function assert_no_match(s, sub, msg)
    if s:find(sub, 1, true) then
        error((msg or "") .. " did not expect '" .. sub .. "' in '" .. s .. "'")
    end
end
local function assert_eq(actual, expected, msg)
    if actual ~= expected then
        error((msg or "assertion failed")
            .. " (expected " .. tostring(expected)
            .. ", got " .. tostring(actual) .. ")")
    end
end

-- ── cell() ──────────────────────────────────────────────────────────

test("cell emits a span with the value", function()
    local s = inline_edit.cell({ value = "Conveyor #3", edit_url = "/x/edit" })
    assert_match(s, '<span')
    assert_match(s, 'Conveyor #3')
end)

test("cell uses outerHTML swap", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/x/edit" })
    assert_match(s, 'hx-swap="outerHTML"')
end)

test("cell is keyboard-activatable (role=button, tabindex)", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/x/edit" })
    assert_match(s, 'role="button"')
    assert_match(s, 'tabindex="0"')
end)

test("cell triggers on click, Enter, Space", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/x/edit" })
    assert_match(s, 'click')
    assert_match(s, "Enter")
    -- Space — the literal char appears in the trigger
    assert_match(s, "key==' '")
end)

test("cell emits hx-get pointing at edit_url", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/assets/42/name/edit" })
    assert_match(s, 'hx-get="/assets/42/name/edit"')
end)

test("cell default label is Edit", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/x/edit" })
    assert_match(s, 'aria-label="Edit"')
    assert_match(s, 'title="Edit"')
end)

test("cell custom label appears in aria-label and title", function()
    local s = inline_edit.cell({ value = "x", edit_url = "/x/edit", label = "Edit asset name" })
    assert_match(s, 'aria-label="Edit asset name"')
    assert_match(s, 'title="Edit asset name"')
end)

test("cell escapes value (no <script> escape)", function()
    local s = inline_edit.cell({ value = "<script>alert(1)</script>", edit_url = "/x/edit" })
    assert_no_match(s, '<script>')
    assert_match(s, '&lt;script&gt;')
end)

test("cell escapes edit_url (no quote-break)", function()
    local s = inline_edit.cell({ value = "x", edit_url = '/x?a="b"' })
    assert_match(s, 'hx-get="/x?a=&quot;b&quot;"')
end)

-- ── editor() ────────────────────────────────────────────────────────

test("editor emits a form with hx-patch", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, '<form')
    assert_match(s, 'hx-patch="/x"')
end)

test("editor uses outerHTML swap targeting itself", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, 'hx-target="this"')
    assert_match(s, 'hx-swap="outerHTML"')
end)

test("editor input has default name=value", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, 'name="value"')
end)

test("editor input takes custom name", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view", name = "title",
    })
    assert_match(s, 'name="title"')
end)

test("editor input pre-fills value attribute", function()
    local s = inline_edit.editor({
        value = "Conveyor #3", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, 'value="Conveyor #3"')
end)

test("editor default labels are Save / Cancel", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, '>Save</button>')
    assert_match(s, '>Cancel</button>')
end)

test("editor custom button labels", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
        save_label = "Update", cancel_label = "Keep",
    })
    assert_match(s, '>Update</button>')
    assert_match(s, '>Keep</button>')
end)

test("editor Cancel button hx-get points at cancel_url", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    -- The cancel button has hx-get="/x/view"
    assert_match(s, 'hx-get="/x/view"')
end)

test("editor wires Esc-to-cancel via hx-trigger", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, "Escape")
end)

test("editor escapes value (XSS in pre-fill)", function()
    local s = inline_edit.editor({
        value = '"><script>alert(1)</script>',
        save_url = "/x", cancel_url = "/x/view",
    })
    assert_no_match(s, '<script>')
    assert_no_match(s, '"><script>')
    -- Value attr should be quote-escaped
    assert_match(s, 'value="&quot;&gt;')
end)

test("editor input has aria-label (default 'Edit value')", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
    })
    assert_match(s, 'aria-label="Edit value"')
end)

test("editor custom aria-label", function()
    local s = inline_edit.editor({
        value = "x", save_url = "/x", cancel_url = "/x/view",
        label = "Asset name",
    })
    assert_match(s, 'aria-label="Asset name"')
end)

-- ── Edge cases ─────────────────────────────────────────────────────

test("cell with omitted value renders empty span (no 'nil' text)", function()
    local s = inline_edit.cell({ edit_url = "/x/edit" })
    -- The display text should be empty, NOT the literal "nil".
    if s:find(">nil<", 1, true) then
        error("nil collapsed to literal 'nil' instead of empty")
    end
    assert_match(s, '<span')
    assert_match(s, '></span>')  -- empty body
end)

test("cell with table-typed value raises a clear error", function()
    local ok, err = pcall(inline_edit.cell, { value = {}, edit_url = "/x/edit" })
    assert_eq(ok, false, "should have errored")
    if not tostring(err):find("must be a string or number", 1, true) then
        error("expected type-narrowing error, got: " .. tostring(err))
    end
end)

test("editor with table-typed save_url raises a clear error", function()
    local ok, err = pcall(inline_edit.editor, {
        value = "x", save_url = {}, cancel_url = "/x/view",
    })
    assert_eq(ok, false, "should have errored")
    if not tostring(err):find("must be a string or number", 1, true) then
        error("expected type-narrowing error, got: " .. tostring(err))
    end
end)

print(string.format("\n%d/%d inline-edit tests passed", pass, pass + fail))
if fail > 0 then os.exit(1) end
